core/watcher/eventsource/value.go

```diff
@@ -4,11 +4,18 @@
 package eventsource
 
 import (
+	"context"
+
 	"gopkg.in/tomb.v2"
 
 	"github.com/juju/juju/core/changestream"
 )
 
+// ValuePredicate is a function that determines whether a change event
+// should be sent to the watcher.
+// Returning false will prevent the events from being sent.
+type ValuePredicate func(context.Context, []changestream.ChangeEvent) bool
```

Comment:
I'm worried about the life-cycle of things closed over by such predicates.

If they can become unusable, but are not coupled to the watcher life-cycle when they should be, there is an avenue for all sorts of subtle bugs.

I understand that this is what the context is supposed to be used for, but what about things explicitly _not_ tied to the watcher's life-cycle, like a service connected to a different DB?

Reply:
I've addressed most of the concerns here.

  1. We return an error from the predicate function. This will tie the predicate to the lifecycle of the watcher.
  2. Pass the TxnRunner to the predicate so users of the predicate don't need to compose the runner from other sources. This should prevent a service connected to a different DB.

Understandably we can't solve people doing bad things, but I think I've solved most of the foot guns.